synapse/push/bulk_push_rule_evaluator.py

```diff
@@ -390,34 +447,19 @@ async def _update_rules_with_member_event_ids(
         if logger.isEnabledFor(logging.DEBUG):
             logger.debug("Found members %r: %r", self.room_id, members.values())
 
-        interested_in_user_ids = {
+        user_ids = {
             user_id
             for user_id, membership in members.values()
             if membership == Membership.JOIN
         }
 
-        logger.debug("Joined: %r", interested_in_user_ids)
-
-        if_users_with_pushers = await self.store.get_if_users_have_pushers(
-            interested_in_user_ids, on_invalidate=self.invalidate_all_cb
-        )
-
-        user_ids = {
-            uid for uid, have_pusher in if_users_with_pushers.items() if have_pusher
-        }
-
-        logger.debug("With pushers: %r", user_ids)
-
-        users_with_receipts = await self.store.get_users_with_read_receipts_in_room(
-            self.room_id, on_invalidate=self.invalidate_all_cb
-        )
-
-        logger.debug("With receipts: %r", users_with_receipts)
+        logger.debug("Joined: %r", user_ids)
 
-        # any users with pushers must be ours: they have pushers
-        for uid in users_with_receipts:
-            if uid in interested_in_user_ids:
-                user_ids.add(uid)
+        # Previously we only considered users with pushers and read receipts in that
+        # room. We can't do this anymore because we use push actions to calculate unread
+        # counts, which don't rely on the user having pushers or sent a read receipt into
+        # the room. Therefore we just need to filter for local users here.
+        user_ids = list(filter(self.is_mine_id, user_ids))
```

Comment:
It's not as much looking up push rules for all users as looking up local members in the room. Currently, if a user is ignored because it has neither any pusher nor a read receipt, Synapse will just ignore them when storing push actions (and as a result, in `action_for_event_by_user`, `rules_by_user` won't include them).
I'm not entirely sure this makes sense for notifications, and it doesn't really make sense for unread counts as [MSC2654](https://github.com/matrix-org/matrix-doc/blob/babolivier/unread_counts/proposals/2654-unread-counts.md) defines what to do if a user joined the room but didn't send a read receipt in it (Synapse doesn't send a read receipt upon joining a room).

Reply:
Though my comment is telling lies here:

> Previously we only considered users with pushers and read receipts

This "and" should be an "or".